# KIVA ANALYTICS
The kiva datasets contain data on money funded to borrowers from different countries.
A lot of information can be deduced from the datasets.

In [1]:
# Importing the packages
import numpy as np
import pandas as pd

### Taking a look at the main dataset

In [3]:
# Read the kiva_loans.csv file
kiva_main_dataset = pd.read_csv('F:/isaka/python4DS/kiva_loans.csv')

# Display the first 5 entries
kiva_main_dataset.head()

id  funded_amount  loan_amount             activity          sector  \
0  653051          300.0        300.0  Fruits & Vegetables            Food   
1  653053          575.0        575.0             Rickshaw  Transportation   
2  653068          150.0        150.0       Transportation  Transportation   
3  653063          200.0        200.0           Embroidery            Arts   
4  653084          400.0        400.0           Milk Sales            Food   

                                                 use country_code   country  \
0            To buy seasonal, fresh fruits to sell.            PK  Pakistan   
1  to repair and maintain the auto rickshaw used ...           PK  Pakistan   
2  To repair their old cycle-van and buy another ...           IN     India   
3  to purchase an embroidery machine and a variet...           PK  Pakistan   
4                           to purchase one buffalo.           PK  Pakistan   

         region currency  partner_id                posted_time  \
0        Lahore      PKR       247.0  2014-01-01 06:12:39+00:00   
1        Lahore      PKR       247.0  2014-01-01 06:51:08+00:00   
2     Maynaguri      INR       334.0  2014-01-01 09:58:07+00:00   
3        Lahore      PKR       247.0  2014-01-01 08:03:11+00:00   
4  Abdul Hakeem      PKR       245.0  2014-01-01 11:53:19+00:00   

              disbursed_time                funded_time  term_in_months  \
0  2013-12-17 08:00:00+00:00  2014-01-02 10:06:32+00:00            12.0   
1  2013-12-17 08:00:00+00:00  2014-01-02 09:17:23+00:00            11.0   
2  2013-12-17 08:00:00+00:00  2014-01-01 16:01:36+00:00            43.0   
3  2013-12-24 08:00:00+00:00  2014-01-01 13:00:00+00:00            11.0   
4  2013-12-17 08:00:00+00:00  2014-01-01 19:18:51+00:00            14.0   

   lender_count                          tags borrower_genders  \
0            12                           NaN           female   
1            14                           NaN   female, female   
2             6  user_favorite, user_favorite           female   
3             8                           NaN           female   
4            16                           NaN           female   

  repayment_interval        date  
0          irregular  2014-01-01  
1          irregular  2014-01-01  
2             bullet  2014-01-01  
3          irregular  2014-01-01  
4            monthly  2014-01-01

## Handling Missing Data
The very first thing before starting on any analytics is to take care of all missing data in the relevant columns(key interest points) by either dropping them or filling them in.<br><br>
The main columns with missing data are disbursed_time column, funded_time column and borrower_genders column. <br><br>
Dropping these columns would lead to loss of other valuable data and filling them in with other values would lead to misguided(false) conclusions, hence we are going to work with only the available data.

In [13]:
# Return all columns and rows without a missing value
kiva_available_data = kiva_main_dataset[kiva_main_dataset.notnull()]

kiva_available_data.head()

id  funded_amount  loan_amount             activity          sector  \
0  653051          300.0        300.0  Fruits & Vegetables            Food   
1  653053          575.0        575.0             Rickshaw  Transportation   
2  653068          150.0        150.0       Transportation  Transportation   
3  653063          200.0        200.0           Embroidery            Arts   
4  653084          400.0        400.0           Milk Sales            Food   

                                                 use country_code   country  \
0            To buy seasonal, fresh fruits to sell.            PK  Pakistan   
1  to repair and maintain the auto rickshaw used ...           PK  Pakistan   
2  To repair their old cycle-van and buy another ...           IN     India   
3  to purchase an embroidery machine and a variet...           PK  Pakistan   
4                           to purchase one buffalo.           PK  Pakistan   

         region currency  partner_id                posted_time  \
0        Lahore      PKR       247.0  2014-01-01 06:12:39+00:00   
1        Lahore      PKR       247.0  2014-01-01 06:51:08+00:00   
2     Maynaguri      INR       334.0  2014-01-01 09:58:07+00:00   
3        Lahore      PKR       247.0  2014-01-01 08:03:11+00:00   
4  Abdul Hakeem      PKR       245.0  2014-01-01 11:53:19+00:00   

              disbursed_time                funded_time  term_in_months  \
0  2013-12-17 08:00:00+00:00  2014-01-02 10:06:32+00:00            12.0   
1  2013-12-17 08:00:00+00:00  2014-01-02 09:17:23+00:00            11.0   
2  2013-12-17 08:00:00+00:00  2014-01-01 16:01:36+00:00            43.0   
3  2013-12-24 08:00:00+00:00  2014-01-01 13:00:00+00:00            11.0   
4  2013-12-17 08:00:00+00:00  2014-01-01 19:18:51+00:00            14.0   

   lender_count                          tags borrower_genders  \
0            12                           NaN           female   
1            14                           NaN   female, female   
2             6  user_favorite, user_favorite           female   
3             8                           NaN           female   
4            16                           NaN           female   

  repayment_interval        date  
0          irregular  2014-01-01  
1          irregular  2014-01-01  
2             bullet  2014-01-01  
3          irregular  2014-01-01  
4            monthly  2014-01-01

## Basic derivations
Now we have a clean dataset without any missing values. The first thing is to look at the basic derivations/computations of the dataset

In [5]:
# Return the basic derivations of the dataset
kiva_available_data.describe()

id  funded_amount    loan_amount     partner_id  \
count  6.712050e+05  671205.000000  671205.000000  657698.000000   
mean   9.932486e+05     785.995061     842.397107     178.199616   
std    1.966113e+05    1130.398941    1198.660073      94.247581   
min    6.530470e+05       0.000000      25.000000       9.000000   
25%    8.230720e+05     250.000000     275.000000     126.000000   
50%    9.927800e+05     450.000000     500.000000     145.000000   
75%    1.163653e+06     900.000000    1000.000000     204.000000   
max    1.340339e+06  100000.000000  100000.000000     536.000000   

       term_in_months   lender_count  
count   671205.000000  671205.000000  
mean        13.739022      20.590922  
std          8.598919      28.459551  
min          1.000000       0.000000  
25%          8.000000       7.000000  
50%         13.000000      13.000000  
75%         14.000000      24.000000  
max        158.000000    2986.000000

### From the above table, we can see that:
1. A total of **671,205** loans were given out during that specific period.
2. An average of **842.40 USD** was disbursed to the borrowers by the field agent.
3. An average of **786.00 USD** was disbursed to the field agent by Kiva.
4. The loans were disburded for an average duration of **14 months**.
5. The minimum loan amount disbursed by Kiva to the field agent was **0.00 USD** while <br> while the maximum amount was **100000.00 USD**.
6. The minimum loan amount disbursed by the field agent to a borrower was **25.00 USD** while <br> while the maximum amount was **100,000.00 USD**. 
7. The  minimum duration for which a loan was disbursed was **1 month** while the maximum duration was **158 months**.
8. There were circumstance(s) when no lender contributed to a certain loan (0 lender count) while the maximum number of lenders a particular loan received was **2986 lenders**.

## Loans borrowed and disbursed per country
To achieve this, we create a pivot table with the countries as the indices and a total summation of the funded and loaned amounts as the values.

In [6]:
# Create a per_country pivot table
loans_per_country = pd.pivot_table(kiva_available_data, index=['country'], values=['funded_amount', 'loan_amount'], aggfunc=[np.sum])

loans_per_country.head(10)

sum            
             funded_amount loan_amount
country                               
Afghanistan        14000.0     14000.0
Albania          2490000.0   2666500.0
Armenia         11186675.0  12915400.0
Azerbaijan       2699575.0   2888700.0
Belize            114025.0    114025.0
Benin             516825.0    518950.0
Bhutan             15625.0     20000.0
Bolivia         18276200.0  19843250.0
Brazil            661025.0    662200.0
Burkina Faso     2909975.0   2972700.0

In [10]:
# Formatting to a more clear dataframe
loans_per_country_summation = loans_per_country['sum']

loans_per_country_summation.head()

funded_amount  loan_amount
country                                
Afghanistan        14000.0      14000.0
Albania          2490000.0    2666500.0
Armenia         11186675.0   12915400.0
Azerbaijan       2699575.0    2888700.0
Belize            114025.0     114025.0

Now we have a clear DataFrame with the countries as indices and the summations of loaned and funded amounts as values. <br><br>
Separating the funded and loan amount columns into Series objects for easier manipulation:

In [27]:
# Selecting the individual summation columns
funded_amts = loans_per_country_summation['funded_amount']
loan_amts = loans_per_country_summation['loan_amount']

Next we select the countries at minimum and maximum values.

In [30]:
# Countries at minimum and maximum funded amount
min_funded_amt_country = funded_amts[(funded_amts == funded_amts.min())]
max_funded_amt_country = funded_amts[(funded_amts == funded_amts.max())]

# Countries at minimum and maximum loan amount
min_loan_amt_country = loan_amts[(loan_amts == loan_amts.min())]
max_loan_amt_country = loan_amts[(loan_amts == loan_amts.max())]

print("Country at minimum funded amount: \n", min_funded_amt_country)
print("Country at maximum funded amount: \n", max_funded_amt_country)
print("Country at minimum loan amount: \n", min_loan_amt_country)
print("Country at maximum loan amount: \n", max_loan_amt_country)

Country at minimum funded amount: 
 country
Virgin Islands    0.0
Name: funded_amount, dtype: float64
Country at maximum funded amount: 
 country
Philippines    54476375.0
Name: funded_amount, dtype: float64
Country at minimum loan amount: 
 country
Guam    4300.0
Name: loan_amount, dtype: float64
Country at maximum loan amount: 
 country
Philippines    55342225.0
Name: loan_amount, dtype: float64


## From the above per country analysis, we can see that:
1. The country that received the largest amount of loan for all sectors was **Philippines**, receiving <br> a total of **55,342,225.00 USD**.
2. The country that received the smallest amount of loan for all sectors was **Guam**, receiving <br> a total of **4,300.00 USD**.

# Loans borrowed per sector across all countries
To achieve this, we create a pivot table with the sectors as the indices and the summation of <br> 
loan amounts as the column.

In [35]:
# return loan amount totals per sector
loans_per_sector = pd.pivot_table(kiva_available_data, index=['sector'], values=['loan_amount'], aggfunc=[np.sum])
loans_per_sector_summation = loans_per_sector['sum']
loans_per_sector_summation

loan_amount
sector                     
Agriculture     143067875.0
Arts             12240325.0
Clothing         37300925.0
Construction      6695150.0
Education        30968525.0
Entertainment     1389150.0
Food            121606150.0
Health            9843400.0
Housing          23663950.0
Manufacturing     5447900.0
Personal Use     14955350.0
Retail           98122900.0
Services         48057450.0
Transportation   11066900.0
Wholesale          995200.0

### Basic analysis of the sectors

In [36]:
# Return the basic sector analytics
loans_per_sector_summation.describe()

loan_amount
count  1.500000e+01
mean   3.769474e+07
std    4.587355e+07
min    9.952000e+05
25%    8.269275e+06
50%    1.495535e+07
75%    4.267919e+07
max    1.430679e+08

## From the above table we see that:
1. The borrowed loans were used in a total of **15** different sectors across all countries.
2. The average borrowed loan amount used per sector in all countries was **37,694,740 USD**.
3. The minimum amount used in a particular sector was **995,200 USD** while the maximum amount was **143,067,900 USD**.

### Additional per-sector analysis
Here we return the sectors with the minimum and maximum amounts of loan used on them, respectively 

In [37]:
# return a series object of all loan amount totals per sector.
loan_amts_per_sector = loans_per_sector_summation['loan_amount']
loan_amts_per_sector

sector
Agriculture       143067875.0
Arts               12240325.0
Clothing           37300925.0
Construction        6695150.0
Education          30968525.0
Entertainment       1389150.0
Food              121606150.0
Health              9843400.0
Housing            23663950.0
Manufacturing       5447900.0
Personal Use       14955350.0
Retail             98122900.0
Services           48057450.0
Transportation     11066900.0
Wholesale            995200.0
Name: loan_amount, dtype: float64

In [38]:
# Return sector with minimum amount
loan_amts_per_sector[loan_amts_per_sector == loan_amts_per_sector.min()]

sector
Wholesale    995200.0
Name: loan_amount, dtype: float64

In [39]:
# Return sector with maximum amount
loan_amts_per_sector[loan_amts_per_sector == loan_amts_per_sector.max()]

sector
Agriculture    143067875.0
Name: loan_amount, dtype: float64

## From the above, 
1. Most of the funds borrowed were used on **Agriculture**, with a total of **143,067,900 USD** used on it.
2. The sector with the smallest amount used on it was **Wholesale**, with a total of **995,200 USD** used on it.